File: packages/dsa-py/src/algorithms/sorting/merge_sort.py
"""Merge Sort — O(n log n) en todos los casos. Estable."""
from typing import TypeVar, Callable, Protocol, cast

T = TypeVar("T")
Cmp = TypeVar("Cmp", bound="Comparable")


class Comparable(Protocol):
    def __le__(self, other: object, /) -> bool: ...


def merge_sort(
    arr: list[T],
    key: Callable[[T], Cmp] | None = None,
) -> list[T]:
    if len(arr) <= 1:
        return arr[:]

    mid   = len(arr) // 2
    left  = merge_sort(arr[:mid], key)
    right = merge_sort(arr[mid:], key)
    return _merge(left, right, key)


def _merge(
    left: list[T],
    right: list[T],
    key: Callable[[T], Cmp] | None,
) -> list[T]:
    result: list[T] = []
    i = j = 0
    def val(x: T) -> Cmp:
        return key(x) if key else cast(Cmp, x)

    while i < len(left) and j < len(right):
        if val(left[i]) <= val(right[j]):
            result.append(left[i]); i += 1
        else:
            result.append(right[j]); j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

File: packages/dsa-ts/src/algorithms/sorting/mergeSort.ts
/**
 * Merge Sort — O(n log n) en todos los casos. Estable.
 * Divide y vencerás: divide el array, ordena cada mitad, fusiona.
 */
export function mergeSort<T>(
  arr: T[],
  comparator: (a: T, b: T) => number = (a, b) => (a < b ? -1 : a > b ? 1 : 0),
): T[] {
  if (arr.length <= 1) return arr;

  const mid   = Math.floor(arr.length / 2);
  const left  = mergeSort(arr.slice(0, mid), comparator);
  const right = mergeSort(arr.slice(mid), comparator);

  return merge(left, right, comparator);
}

function merge<T>(left: T[], right: T[], cmp: (a: T, b: T) => number): T[] {
  const result: T[] = [];
  let i = 0, j = 0;

  while (i < left.length && j < right.length) {
    if (cmp(left[i] as T, right[j] as T) <= 0) {
      result.push(left[i++] as T);
    } else {
      result.push(right[j++] as T);
    }
  }
  while (i < left.length)  result.push(left[i++]  as T);
  while (j < right.length) result.push(right[j++] as T);
  return result;
}
